File: 1269-number-of-ways-to-stay-in-the-same-place-after-some-steps/1269-number-of-ways-to-stay-in-the-same-place-after-some-steps.cpp
class Solution {
public:
    int numWays(int steps, int arrLen) {
        int n = min(arrLen, steps / 2 + 2);
        long M = 1e9 + 7;
        vector<long> dp(n);
        dp[0] = 1;
        for(int k = 0; k < steps; ++k) {
             auto dp2 = dp;
            for(int i = 0; i < n; ++i) {
                dp[i] = dp2[i] + (i >= 1 ? dp2[i - 1] : 0) + (i < n - 1 ? dp2[i + 1] : 0);
                dp[i] %= M;
            }
        }
  
        return dp[0];
    }
};

File: 1269-number-of-ways-to-stay-in-the-same-place-after-some-steps/1269-number-of-ways-to-stay-in-the-same-place-after-some-steps.py
class Solution(object):
    def numWays(self, steps, arrLen):
        """
        :type steps: int
        :type arrLen: int
        :rtype: int
        """
        n = min(steps//2+2, arrLen)
        dp = [0] * n
        M = 10**9 + 7
        dp[0] = 1
        for k in range(steps):
            dp2 = dp[:]
            for i in range(n):
                dp[i] = dp2[i] + (dp2[i-1] if i>=1 else 0) + (dp2[i+1] if i+1<n else 0)
                dp[i] %= M
        return dp[0]

                